mmda/predictors/hf_predictors/span_group_classification_predictor.py

```diff
@@ -0,0 +1,358 @@
+from mmda.types.document import Document
+from mmda.predictors.hf_predictors.base_hf_predictor import BaseHFPredictor
+
+
+class SpanGroupClassificationBatch:
+    def __init__(
+            self,
+            input_ids: List[List[int]],
+            attention_mask: List[List[int]],
+            span_group_ids: List[List[Optional[int]]],
+            context_id: List[int]
+    ):
+        assert len(input_ids) == len(attention_mask) == len(span_group_ids) == len(context_id), \
+            f"Inputs to batch arent same length"
+        self.batch_size = len(input_ids)
+        assert [len(example) for example in input_ids] == \
+               [len(example) for example in attention_mask] == \
+               [len(example) for example in span_group_ids], f"Examples in batch arent same length"
+        self.input_ids = input_ids
+        self.attention_mask = attention_mask
+        self.span_group_ids = span_group_ids
+        self.context_id = context_id
+
+
+class SpanGroupClassificationPrediction:
+    def __init__(self, context_id: int, span_group_id: int, label: str, score: float):
+        self.context_id = context_id
+        self.span_group_id = span_group_id
+        self.label = label
+        self.score = score
+
+
+class SpanGroupClassificationPredictor(BaseHFPredictor):
+    """
+    This is a generic wrapper around Huggingface Token Classification models.
+
+    First, we need a `span_group_name` which defines the Document field that we will treat as the
+    target unit of prediction. For example, if `span_group_name` is 'tokens', then we expect
+    to classify every Document.token.  But technically, `span_group_name` could be anything,
+    such as `words` or `rows` or any SpanGroup.
+
+    Second, we need a `context_name` which defines the Document field that we will treat as the
+    intuitive notion of an "example" that we want to run our model over. For example, if
+    `context_name` is 'pages', then we'll loop over each page, running our classifier
+    over all the 'tokens' in each page. If the `context_name` is `bib_entries`, then we'll
+    loop over each bib entry, running our classifier over the 'tokens' in each page.
+
+    The key consequence of defining a `context_name` is, when the model constructs batches
+    of sequences that fit within the Huggingface transformer's window, it will *not*
+    mix sequences from different contexts into the same batch.
+
+    @kylel
+    """
+
+    REQUIRED_BACKENDS = ["transformers", "torch", "smashed"]
+    REQUIRED_DOCUMENT_FIELDS = []
+
+    _SPAN_GROUP = 'inputs'
+    _CONTEXT_ID = 'context_id'
+
+    _HF_RESERVED_INPUT_IDS = 'input_ids'
+    _HF_RESERVED_ATTN_MASK = 'attention_mask'
+    _HF_RESERVED_WORD_IDS = 'word_ids'
+    _HF_RESERVED_WORD_PAD_VALUE = -1
+
+    def __init__(
+            self,
+            model: Any,
+            config: Any,
+            tokenizer: Any,
+            span_group_name: str,
+            context_name: str,
```

Comment:
not something to fix, but we really ought to come up with a standard name for contexts/annotations/groups/???  It took me way to long to figure out what this was